File: Qual/3_22_2/32_22_2_QE_prob2.py
class GNode:
    def __init__(self, id):
        self.id = id
    def __str__(self):
        return self.id
    
a, b, c, d = GNode('a'), GNode('b'), GNode('c'), GNode('d')
G = dict()
G[a], G[b], G[c], G[d] = [b, c], [d], [], [c]

def paths(Gd : dict, s:GNode, t:GNode)->list:
    ans = []
    tmp = []
    def backtracking(Gd, st, tgt):
        if (st == tgt):
            ans.append(tmp.copy()) #tmp로 넣어주면 tmp가 계속 바뀌어서 ans에도 영향을 줌
            return
        for adj in G[st]:
            tmp.append(adj.id)
            backtracking(Gd, adj, tgt)
            tmp.remove(adj.id)
    tmp.append(s.id)
    backtracking(Gd, s, t)
    return ans

if __name__ == "__main__":
    print(paths(G, a, c))

File: Qual/3_22_2/32_22_2_QE_prob2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

class GNode {
public:
    std::string id;
    GNode(std::string id) : id(id) {}
};

typedef std::unordered_map<GNode*, std::vector<GNode*>> Graph;

void backtracking(const Graph& G, GNode* st, GNode* tgt, std::vector<std::string>& tmp, std::vector<std::vector<std::string>>& ans) {
    if (st == tgt) {
        ans.push_back(tmp);
        return;
    }
    for (GNode* adj : G.at(st)) {
        tmp.push_back(adj->id);
        backtracking(G, adj, tgt, tmp, ans);
        tmp.pop_back();
    }
}

std::vector<std::vector<std::string>> paths(const Graph& G, GNode* s, GNode* t) {
    std::vector<std::vector<std::string>> ans;
    std::vector<std::string> tmp;
    tmp.push_back(s->id);
    backtracking(G, s, t, tmp, ans);
    return ans;
}

int main() {
    GNode a("a"), b("b"), c("c"), d("d");
    Graph G;
    G[&a] = {&b, &c};
    G[&b] = {&d};
    G[&c] = {};
    G[&d] = {&c};

    std::vector<std::vector<std::string>> result = paths(G, &a, &c);

    for (const auto& path : result) {
        for (const auto& node : path) {
            std::cout << node << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}
